cmd/thanos/receive.go

```diff
@@ -140,56 +140,96 @@ func runReceive(
 		MinBlockDuration:  model.Duration(time.Hour * 2),
 		MaxBlockDuration:  model.Duration(time.Hour * 2),
 	}
+	db := receive.NewFlushableStorage(
+		dataDir,
+		log.With(logger, "component", "tsdb"),
+		reg,
+		tsdbCfg,
+	)
 
 	localStorage := &tsdb.ReadyStorage{}
-	receiver := receive.NewWriter(log.With(logger, "component", "receive-writer"), localStorage)
 	webHandler := receive.NewHandler(log.With(logger, "component", "receive-handler"), &receive.Options{
-		Receiver:          receiver,
 		ListenAddress:     remoteWriteAddress,
 		Registry:          reg,
-		ReadyStorage:      localStorage,
 		Endpoint:          endpoint,
 		TenantHeader:      tenantHeader,
 		ReplicaHeader:     replicaHeader,
 		ReplicationFactor: replicationFactor,
 	})
 
+	confContentYaml, err := objStoreConfig.Content()
+	if err != nil {
+		return err
+	}
+	upload := true
+	if len(confContentYaml) == 0 {
+		level.Info(logger).Log("msg", "No supported bucket was configured, uploads will be disabled")
+		upload = false
+	}
+
 	// Start all components while we wait for TSDB to open but only load
 	// initial config and mark ourselves as ready after it completed.
-	dbOpen := make(chan struct{})
+
+	// dbReady signals when TSDB is ready and the Store gRPC server can start.
+	dbReady := make(chan struct{})
+	// updateDB signals when TSDB needs to be flushed and updated.
+	updateDB := make(chan struct{})
+	// uploadC signals when new blocks should be uploaded.
+	uploadC := make(chan struct{})
+	// uploadDone signals when uploading has finished.
+	uploadDone := make(chan struct{})
 	level.Debug(logger).Log("msg", "setting up tsdb")
 	{
 		// TSDB.
 		cancel := make(chan struct{})
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting TSDB ...")
-				db, err := tsdb.Open(
-					dataDir,
-					log.With(logger, "component", "tsdb"),
-					reg,
-					tsdbCfg,
-				)
-				if err != nil {
-					close(dbOpen)
-					return fmt.Errorf("opening storage failed: %s", err)
-				}
-				level.Info(logger).Log("msg", "tsdb started")
-
-				startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
-				localStorage.Set(db, startTimeMargin)
-				webHandler.StorageReady()
-				level.Info(logger).Log("msg", "server is ready to receive web requests.")
-				close(dbOpen)
-				<-cancel
-				return nil
-			},
-			func(err error) {
-				if err := localStorage.Close(); err != nil {
-					level.Error(logger).Log("msg", "error stopping storage", "err", err)
+		// Before actually starting, we need to make sure
+		// the WAL is flushed.
```

Comment:
Just for correctness; if for some reason the last time the receiver ran it quit unexpectedly and was not able to flush the WAL on shutdown, we want to make sure it is flushed so that the new blocks don't have mixed data.